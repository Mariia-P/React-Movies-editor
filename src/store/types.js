export const LIKED_MOVIE = 'LIKED_MOVIE';

export const CHANGE_RATING = 'CHANGE_RATING';

export const MOVIE_LIST_LOADED = 'MOVIE_LIST_LOADED';
export const MOVIE_LIST_LOADING = 'MOVIE_LIST_LOADING';
export const MOVIE_LIST_LOAD_ERROR = 'MOVIE_LIST_LOAD_ERROR';

export const SORT_MOVIES = 'SORT_MOVIES';

export const CHANGE_SEARCH = 'CHANGE_SEARCH';
export const RESET_SEARCH = 'RESET_SEARCH';

export const SELECT_MOVIE = 'SELECT_MOVIE';
