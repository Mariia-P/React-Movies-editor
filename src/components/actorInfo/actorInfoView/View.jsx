import React from 'react';
import PT from 'prop-types';

import styles from '../ActorInfo.module.scss';

const View = ({ selectedActor, textContent }) => {
  const { name, description, thumbnail } = selectedActor;

  return (
    <>
      <div className={styles.actor__basics}>
        <div>
          <div className={styles.actor__info_name}>{name}</div>

        </div>
        <img className={styles.actor__poster} src={thumbnail} alt={name} />
      </div>
      <div className={styles.actor__descr}>
        <p>{textContent[0]}</p>
        {description}
      </div>

    </>
  );
};

View.propTypes = {
  selectedActor: PT.shape({
    name: PT.string,
    thumbnail: PT.string,
    description: PT.string,
  }),
  textContent: PT.arrayOf(PT.string).isRequired,
};

View.defaultProps = {
  selectedActor: {
    name: '',
    description: '',
    thumbnail: '',
  },
};

export default View;
