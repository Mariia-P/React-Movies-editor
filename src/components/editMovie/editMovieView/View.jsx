import React from 'react';
import { useFormik } from 'formik';
import * as Yup from 'yup';
import PT from 'prop-types';
import cnBind from 'classnames/bind';

import { useNavigate } from 'react-router-dom';
import useMoviesProcessor from '../../utils/hooks/useMoviesProcessor';
import styles from '../EditMovie.module.scss';

const cx = cnBind.bind(styles);

const View = ({ movie, textContent }) => {
  const navigate = useNavigate();

  const { onSubmitEditMovieForm } = useMoviesProcessor();

  const formik = useFormik({
    initialValues: {
      name: '',
      thumbnail: '',
      directorName: '',
      description: '',
    },
    validationSchema: Yup.object({
      name: Yup.string()
        .min(2, 'at least 2 characters!')
        .max(50, 'maximum 50 characters')
        .required('This is a required field!'),
      thumbnail: Yup.string()
        .matches(/^(http|https)[\W\w]+(.jpg|.png|.jpeg)$/g, 'enter the correct link to the picture'),
      directorName: Yup.string()
        .min(2, 'at least 2 characters!')
        .max(50, 'maximum 50 characters'),
      description: Yup.string()
        .min(10, 'at least 10 characters!')
        .max(1000, 'maximum 1000 characters'),

    }),
    onSubmit: (values) => {
      const changedFields = {};
      Object.entries(values).forEach(([key, value]) => {
        if (value) {
          changedFields[key] = value;
        }
      });
      onSubmitEditMovieForm(movie, changedFields);
      navigate('/');
    },
  });

  const handleBackClick = (e) => {
    e.preventDefault();
    navigate(-1);
  };

  const createInputClasses = (tagName, inputName) => cx({
    [tagName]: true,
    error: !!(formik.errors[inputName] && formik.touched[inputName]),
  });

  return (
    <form className={styles.form} action="url" onSubmit={formik.handleSubmit}>
      <input
        value={formik.values.name}
        type="text"
        placeholder={textContent[0]}
        required
        name="name"
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        className={createInputClasses('input', 'name')}
      />
      {formik.errors.name && formik.touched.name ? <div>{formik.errors.name}</div> : null}

      <input
        value={formik.values.thumbnail}
        type="text"
        placeholder={textContent[1]}
        name="thumbnail"
        autoComplete="off"
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        className={createInputClasses('input', 'thumbnail')}
      />
      {formik.errors.thumbnail && formik.touched.thumbnail
        ? <div>{formik.errors.thumbnail}</div> : null}

      <input
        value={formik.values.directorName}
        type="text"
        placeholder={textContent[2]}
        name="directorName"
        autoComplete="off"
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        className={createInputClasses('input', 'directorName')}
      />
      {formik.errors.directorName && formik.touched.directorName
        ? <div>{formik.errors.directorName}</div> : null}

      <div className="form__textarea-wrepper">
        <p>{textContent[3]}</p>
        <textarea
          value={formik.values.description}
          name="description"
          id="txt"
          cols="43"
          rows="10"
          onChange={formik.handleChange}
          onBlur={formik.handleBlur}
          className={createInputClasses('textarea', 'description')}
        />
        {formik.errors.description && formik.touched.description
          ? <div>{formik.errors.description}</div> : null}

      </div>
      <div className={styles.form__button_wrapper}>
        <button
          type="submit"
          className="button button__main"
        >
          <div className="inner">{textContent[4]}</div>
        </button>

        <button
          type="button"
          className="button button__secondary"
          onClick={handleBackClick}
        >
          <div className="inner">{textContent[5]}</div>
        </button>
      </div>
    </form>

  );
};

View.propTypes = {
  movie: PT.shape({
    id: PT.oneOfType([
      PT.string,
      PT.number,
    ]),
    name: PT.string,
    actors: PT.arrayOf(PT.shape({
      id: PT.number,
      name: PT.string,
    })),
    directorName: PT.string,
    rating: PT.number,
    likes: PT.number,
    thumbnail: PT.string,
    description: PT.string,
  }),
  textContent: PT.arrayOf(PT.string).isRequired,
};

View.defaultProps = {
  movie: {
    id: null,
    name: '',
    description: '',
    thumbnail: '',
    directorName: '',
    likes: 0,
    rating: 0,
    actors: [],
  },
};

export default View;
