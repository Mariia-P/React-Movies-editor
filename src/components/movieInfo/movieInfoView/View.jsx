/* eslint-disable no-shadow */
import React from 'react';
import PT from 'prop-types';
import { Link, useNavigate } from 'react-router-dom';

import useMoviesProcessor from '../../utils/hooks/useMoviesProcessor';
import Rating from '../../rating/Rating';
import styles from '../MovieInfo.module.scss';

const View = ({ selectedMovie, textContent }) => {
  const navigate = useNavigate();

  const { onDelete } = useMoviesProcessor();
  const {
    name, description, thumbnail, rating, likes, directorName, actors, id,
  } = selectedMovie;

  const onMainClick = (e) => {
    e.preventDefault();
    navigate('/');
  };

  const onEditClick = (e) => {
    e.preventDefault();
    navigate(`/edit/${id}`);
  };

  const actorsButtons = actors.map(({ id, actorName }) => (
    <Link
      to={`/actor/${id}`}
      className={styles.movie__actor_link}
      key={id}
    >
      {actorName}
    </Link>
  ));

  return (
    <>
      <div className={styles.movie__basics}>
        <div>
          <div className={styles.movie__info_name}>{name}</div>
          <div className={styles.movie__like}>
            {textContent[0]}
            {likes}
          </div>
          <Rating rating={rating} />
          <div className={styles.movie__button_wrapper}>
            <button
              className="button button__main"
              type="button"
              onClick={(e) => { onDelete(id); onMainClick(e); }}
            >
              <div className="inner">{textContent[1]}</div>
            </button>
            <button
              className="button button__secondary"
              type="button"
              onClick={(e) => { onEditClick(e); }}
            >
              <div className="inner">{textContent[2]}</div>
            </button>
          </div>

        </div>
        <img className={styles.movie__poster} src={thumbnail} alt={name} />
      </div>
      <div className={styles.movie__descr}>
        <p>{textContent[3]}</p>
        {directorName}
      </div>
      <div className={styles.movie__descr}>
        <p>{textContent[4]}</p>
        {actorsButtons}
      </div>
      <div className={styles.movie__descr}>
        <p>{textContent[5]}</p>
        {description}
      </div>

    </>
  );
};

View.propTypes = {
  selectedMovie: PT.shape({
    id: PT.oneOfType([
      PT.string,
      PT.number,
    ]),
    name: PT.string,
    actors: PT.arrayOf(PT.shape({
      id: PT.number,
      actorName: PT.string,
    })),
    directorName: PT.string,
    rating: PT.number,
    likes: PT.number,
    thumbnail: PT.string,
    description: PT.string,
  }),
  textContent: PT.arrayOf(PT.string).isRequired,
};

View.defaultProps = {
  selectedMovie: {
    id: null,
    name: '',
    description: '',
    thumbnail: '',
    directorName: '',
    likes: 0,
    rating: 0,
    actors: [],
  },
};

export default View;
