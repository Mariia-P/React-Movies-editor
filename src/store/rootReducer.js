import { combineReducers } from 'redux';
import movieListReducer from './reducers/movieListReducer';

export default combineReducers({
  movies: movieListReducer,
});
