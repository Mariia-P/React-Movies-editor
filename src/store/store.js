/* eslint-disable no-underscore-dangle */
import { configureStore } from '@reduxjs/toolkit';
import authReducer from './reducers/authReducer';
import movieListReducer from './reducers/movieListReducer';

// eslint-disable-next-line no-unused-vars
const customMiddleWare = (store) => (next) => (action) => {
  console.log('Action that was triggered:', action.type);

  next(action);
};

// eslint-disable-next-line no-unused-vars
const checkUserAuthMiddleWare = (store) => (next) => (action) => {
  if (action.type === 'auth/registerUser') {
    const users = localStorage.getItem('users')
      ? JSON.parse(localStorage.getItem('users'))
      : [];
    const isAlreadyRegisteredUser = users.some(
      ({ email }) => email === action.payload.email,
    );
    if (isAlreadyRegisteredUser) {
      next({ ...action, type: 'auth/registeredFromWrongPage' });
      return;
    }
    users.push(action.payload);
    localStorage.setItem('users', JSON.stringify(users));
    next({ ...action, type: 'auth/authenticateUser' });
    return;
  }

  if (action.type === 'auth/loginUser') {
    const users = localStorage.getItem('users')
      ? JSON.parse(localStorage.getItem('users'))
      : [];
    const isAlreadyRegisteredUser = users.some(
      ({ email }) => email === action.payload.email,
    );
    if (!isAlreadyRegisteredUser) {
      next({ ...action, type: 'auth/loginFromWrongPage' });
      return;
    }
    next({ ...action, type: 'auth/authenticateUser' });
    return;
  }

  next(action);
};

export default configureStore({
  reducer: {
    movies: movieListReducer,
    auth: authReducer,
  },
  devTools: process.env.NODE_ENV !== 'production',
  middleware: (getDefaultMiddleware) => getDefaultMiddleware()
    .concat(customMiddleWare, checkUserAuthMiddleWare),
});
