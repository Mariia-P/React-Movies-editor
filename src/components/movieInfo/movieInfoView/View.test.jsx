import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import View from './View';
import store from '../../../store/store';

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
  Link: 'a',
  Route: ({ children, path }) => children({ match: path === '/somewhere' }),
}));

describe('View for MovieInfo', () => {
  const renderWithStore = (
    component,
  ) => ({
    ...render(

      <Provider store={store}>
        {component}
      </Provider>,

    ),
  });

  const textContent = [
    'Likes:',
    'Delete',
    'Edit',
    'Director:',
    'Actors:',
    'Description:',
  ];

  const selectedMovie = {
    id: 25,
    name: 'test',
    description: 'test',
    thumbnail: '',
    directorName: 'test',
    likes: 5,
    rating: 5,
    actors: [{ id: 1, actorName: 'test' }, { id: 2, actorName: 'test' }, {
      id: 3,
      name: 'test',
    }],
  };

  it('should match snapshot', () => {
    const component = renderWithStore(
      <View
        selectedMovie={selectedMovie}
        textContent={textContent}
      />,
      store,
    );

    expect(component).toMatchSnapshot();
  });
});
