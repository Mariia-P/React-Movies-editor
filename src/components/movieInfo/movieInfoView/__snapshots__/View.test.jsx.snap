// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View for MovieInfo should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="movie__basics"
      >
        <div>
          <div
            class="movie__info_name"
          >
            test
          </div>
          <div
            class="movie__like"
          >
            Likes:
            5
          </div>
          <div
            class="rating__wrapper"
          >
            <button
              class="rating__button rating__button_full"
              data-i="1"
              disabled=""
              type="button"
            >
              <img
                alt="star"
                src="Star-1.svg"
              />
            </button>
            <button
              class="rating__button rating__button_full"
              data-i="2"
              disabled=""
              type="button"
            >
              <img
                alt="star"
                src="Star-1.svg"
              />
            </button>
            <button
              class="rating__button rating__button_full"
              data-i="3"
              disabled=""
              type="button"
            >
              <img
                alt="star"
                src="Star-1.svg"
              />
            </button>
            <button
              class="rating__button rating__button_full"
              data-i="4"
              disabled=""
              type="button"
            >
              <img
                alt="star"
                src="Star-1.svg"
              />
            </button>
            <button
              class="rating__button rating__button_full"
              data-i="5"
              disabled=""
              type="button"
            >
              <img
                alt="star"
                src="Star-1.svg"
              />
            </button>
          </div>
          <div
            class="movie__button_wrapper"
          >
            <button
              class="button button__main"
              type="button"
            >
              <div
                class="inner"
              >
                Delete
              </div>
            </button>
            <button
              class="button button__secondary"
              type="button"
            >
              <div
                class="inner"
              >
                Edit
              </div>
            </button>
          </div>
        </div>
        <img
          alt="test"
          class="movie__poster"
          src=""
        />
      </div>
      <div
        class="movie__descr"
      >
        <p>
          Director:
        </p>
        test
      </div>
      <div
        class="movie__descr"
      >
        <p>
          Actors:
        </p>
        <a
          class="movie__actor_link"
          to="/actor/1"
        >
          test
        </a>
        <a
          class="movie__actor_link"
          to="/actor/2"
        >
          test
        </a>
        <a
          class="movie__actor_link"
          to="/actor/3"
        />
      </div>
      <div
        class="movie__descr"
      >
        <p>
          Description:
        </p>
        test
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="movie__basics"
    >
      <div>
        <div
          class="movie__info_name"
        >
          test
        </div>
        <div
          class="movie__like"
        >
          Likes:
          5
        </div>
        <div
          class="rating__wrapper"
        >
          <button
            class="rating__button rating__button_full"
            data-i="1"
            disabled=""
            type="button"
          >
            <img
              alt="star"
              src="Star-1.svg"
            />
          </button>
          <button
            class="rating__button rating__button_full"
            data-i="2"
            disabled=""
            type="button"
          >
            <img
              alt="star"
              src="Star-1.svg"
            />
          </button>
          <button
            class="rating__button rating__button_full"
            data-i="3"
            disabled=""
            type="button"
          >
            <img
              alt="star"
              src="Star-1.svg"
            />
          </button>
          <button
            class="rating__button rating__button_full"
            data-i="4"
            disabled=""
            type="button"
          >
            <img
              alt="star"
              src="Star-1.svg"
            />
          </button>
          <button
            class="rating__button rating__button_full"
            data-i="5"
            disabled=""
            type="button"
          >
            <img
              alt="star"
              src="Star-1.svg"
            />
          </button>
        </div>
        <div
          class="movie__button_wrapper"
        >
          <button
            class="button button__main"
            type="button"
          >
            <div
              class="inner"
            >
              Delete
            </div>
          </button>
          <button
            class="button button__secondary"
            type="button"
          >
            <div
              class="inner"
            >
              Edit
            </div>
          </button>
        </div>
      </div>
      <img
        alt="test"
        class="movie__poster"
        src=""
      />
    </div>
    <div
      class="movie__descr"
    >
      <p>
        Director:
      </p>
      test
    </div>
    <div
      class="movie__descr"
    >
      <p>
        Actors:
      </p>
      <a
        class="movie__actor_link"
        to="/actor/1"
      >
        test
      </a>
      <a
        class="movie__actor_link"
        to="/actor/2"
      >
        test
      </a>
      <a
        class="movie__actor_link"
        to="/actor/3"
      />
    </div>
    <div
      class="movie__descr"
    >
      <p>
        Description:
      </p>
      test
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
